### PR TITLE
Kafka-4351: Topic regex behavioral change with MirrorMaker new consumer

### PR DESCRIPTION
**Pull request by huxihx**
Support CSV values in the regexp for MirrorMaker new consumer as OldConsumer does.

### CODE REVIEW

**Comment by huxihx** [comment]
@joestein please help review this. Thanks.

**Comment by ijuma** [comment]
Thanks for the PR. A couple of things:
1. This could break existing users, so we need to decide if we're OK with that. Probably worth starting a mailing list thread to get wider input. If we go ahead, we should add a note to the upgrade notes so that users are aware of the change.
2. There should be tests to verify the change.

**Comment by huxihx** [comment]
@ijuma, totally agreed on that we need to test the change. As for the doc update thing, I think we might not need to do this since right now the document already claims:

> For convenience we allow the use of ',' instead of '|' to specify a list of topics.

And we know this is exactly what this fix is about. Without the change, only old consumer supports CSV-typed regular expression because of the fact that TopicFilter will convert them internally. But the new consumer is not able to do this. Does it make sense?

**Comment by ijuma** [comment]
That's a fair point, if we already document it like that, then it's just a bug fix. Let's add a note to the upgrade notes still though (`upgrade.html`).

**Comment by huxihx** [comment]
@ijuma Please guide what I should do next. I am a freshman to the process. Thanks.

File: core/src/main/scala/kafka/tools/MirrorMaker.scala
Comment (vahidhashemian) [inline core/src/main/scala/kafka/tools/MirrorMaker.scala]: What if we use something like this

```
          val whitelist = Whitelist(whitelistOpt.get)
          consumer.subscribe(Pattern.compile(whitelist.regex), consumerRebalanceListener)
```

so that regex manipulation is hardcoded in one place?

Comment (huxihx) [reply core/src/main/scala/kafka/tools/MirrorMaker.scala]: @vahidhashemian Agreed on your suggestion. And if we employs Whitelist here,  we need to additionally do a minor change for the exception capturing. Here is the analysis:

Before initializing Whitelist, the construction method for TopicFilter will be executed, which invokes Pattern.compile(regex). If an 'invalid' regex is offered, TopicFilter captures it and throws **RuntimeException**(not PatternSyntaxException), which means, later, the init method for MirrorMakerNewConsumer fails to capture this exception since it stares at PatternSyntaxException (although the impact is ignorable since the enclosing run method captures throwable)

Does it make sense? If you agree with the analysis above, I will make another commit following your suggestion and have this PR ready to be reviewed again. Is it okay to you?

Comment (vahidhashemian) [reply core/src/main/scala/kafka/tools/MirrorMaker.scala]: Thanks for explaining. It makes sense to me.
On a separate note, do you think it makes sense to add some unit test for this change?

Comment (huxihx) [reply core/src/main/scala/kafka/tools/MirrorMaker.scala]: @vahidhashemian Testing is definitely important but as for this case, I am afriad what we need to test  is whether Whitelist behaves as expected, which has already been tested in TopicFilterTest. Besides, I made a new commit following your suggestion, please have a review. Thanks.

Comment (vahidhashemian) [reply core/src/main/scala/kafka/tools/MirrorMaker.scala]: Looks like you committed more than needed in the last try?
On the unit test note, I'm wondering if it makes sense to add (at least) one unit test that verifies the issue raised in the JIRA would not occur with this patch?

Comment (huxihx) [reply core/src/main/scala/kafka/tools/MirrorMaker.scala]: @vahidhashemian  already committed a new one. And as for the unit test thing, which file do you think should be better for the test case to be added in?  TopicFilterTest or MirrorMakerTest?

**Comment by huxihx** [comment]
@vahidhashemian @ijuma any feedback for this PR? Thanks.

File: core/src/main/scala/kafka/tools/MirrorMaker.scala
Comment (vahidhashemian) [reply core/src/main/scala/kafka/tools/MirrorMaker.scala]: For the unit test I'd add it to MirrorMakerTest.

**Comment by huxihx** [comment]
@joestein Could you help review this PR? Thanks.

**Comment by vahidhashemian** [comment]
@amethystic The change you made to fix the issue looks good to me. But @ijuma's concerns above should still be addressed in my opinion.

**Comment by huxihx** [comment]
@vahidhashemian Thanks for your feedback. As for @ijuma 's suggestion, seems that he mentions of upgrade.html, so do you know how to add a update not for this PR? How should I do then ?

**Comment by ijuma** [comment]
`upgrade.html` is in `docs` folder. Not sure what you mean by "a update not for this PR".

**Comment by huxihx** [comment]
@ijuma Thanks. So can this PR be closed now?

**Comment by ijuma** [comment]
@amethystic no, as it has not been merged yet. To clarify: before we merge it, we need:
1. Tests
2. Add a note to `docs/upgrade.html` mentioning this change

**Comment by huxihx** [comment]
@ijuma Thanks for the advice. Already updated the upgrade note. Besides, I did a test in my local environment, which was passed. Both single topic and regular expressions are supported now for the new consumer of MirrorMaker. Do you think we need to add some unit test cases although it is not an easy work to do this?

**Comment by huxihx** [comment]
@ijuma could you please address the comments above? Thanks.

**Comment by ijuma** [comment]
@amethystic, yes it would be good to add at least one test.

**Comment by huxihx** [comment]
@ijuma Sorry, but do you mean the unit test?

**Comment by huxihx** [comment]
@ijuma Seems no MirrorMaker unit test is actually offered, although there is a MirrorMakerTest. Most of the stuffs in the MirrorMaker is private, it's a little hard to test them outside it. Could you give some advice?

**Comment by huxihx** [comment]
@ijuma  added a test case for this PR. Please have a review. Thanks.

File: core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala]: I suggest creating a new test under `integration/kafka/tools/MirrorMaker.scala` instead of reusing the existing unit test to avoid making the existing unit tests slower due to initialisation that they don't need.

Also, it seems like it may be better to inherit from `KafkaServerTestHarness` than `ZooKeeperTestHarness`?

File: core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala]: Instead of setting the serializers like this, you can just pass them to the `KafkaProducer` constructor.

File: core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala]: `close` should have the same effect as `flush`. Does the test fail if you remove `flush`?

File: core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala]: You should probably call `get` on the returned `Future` to ensure that we get a useful error if there is a failure. Also, you can omit `[Array[Byte], Array[Byte]]` in both cases above.

File: core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala]: This comment is redundant.

File: core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala]: Is there a reason why we use a `Properties` instance for the producer and a `HashMap` for the consumer? It would be good to be consistent (either choice is fine).

File: core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala]: Typo in group name.

File: core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala]: Again, you can pass the deserializers via the constructor.

File: core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala]: I thought we wanted to test multiple regular expressions separated by commas?

File: core/src/main/scala/kafka/tools/MirrorMaker.scala
Comment (ijuma) [inline core/src/main/scala/kafka/tools/MirrorMaker.scala]: Why is this method here? Can it not be in the test?

File: core/src/main/scala/kafka/tools/MirrorMaker.scala
Comment (ijuma) [inline core/src/main/scala/kafka/tools/MirrorMaker.scala]: Instead of `whitelistOpt.isDefined`, we can write it as:

```scala
whitelistOpt.foreach { whitelist =>
  try {
    consumer.subscribe(Pattern.compile(Whitelist(whitelist).regex), consumerRebalanceListener)
    ...
  }
}
```

**Review by ijuma — CHANGES_REQUESTED** [review]
Sorry for the delay. Thanks for adding the test, very useful. I left a few suggestions and questions.

**Comment by huxihx** [comment]
@ijuma already addressed most of your comments. Please have a review. Thanks

File: core/src/main/scala/kafka/tools/MirrorMaker.scala
Comment (ijuma) [inline core/src/main/scala/kafka/tools/MirrorMaker.scala]: Should this be `private[tools]`?

File: core/src/main/scala/kafka/tools/MirrorMaker.scala
Comment (ijuma) [inline core/src/main/scala/kafka/tools/MirrorMaker.scala]: Should this be `private[tools]`?

File: core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: We don't need `tearDown` and `setUp` defined here if they just call `super`.

File: core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: Probably left unintentionally?

File: core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: In Scala, `data.topic == topic` is more idiomatic. Same for the next line.

File: core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: No need to catch and rethrow, just let the original exception bubble up.

File: core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: You can set the class itself instead of the name, I think. e.g. `classOf[ByteArraySerializer]`. Same in the line below.

File: core/src/main/scala/kafka/tools/MirrorMaker.scala
Comment (ijuma) [reply core/src/main/scala/kafka/tools/MirrorMaker.scala]: What are your thoughts on this comment?

File: core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: I thought `MirrrorMakerTest` was fine, why did you rename it?

File: core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: As I said before, it would be good to call `get` on the result of `send` to ensure exceptions get bubbled up.

**Review by ijuma — CHANGES_REQUESTED** [review]
Thanks, I think we're close. Left a few more comments.

File: core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala
Comment (huxihx) [reply core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: The producer here is not of type 'KafkaProducer' but 'MirrorMakerProducer'. It's send method is already KafkaProducer.send().get() by default if not explicitly specifying 'async'.  Does it make sense?

Comment (ijuma) [reply core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: Yes, thanks for clarifying.

Comment (ijuma) [reply core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: Hmm, the default seems to be `async`?

```scala
val sync = producerProps.getProperty("producer.type", "async").equals("sync")
```

File: core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala
Comment (huxihx) [reply core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala]: I replaced all "|" with commas. Is that what you expect?

File: core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala
Comment (huxihx) [reply core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: The compiler will complain "MirrorMakerTest is already defined as class MirrorMakerTest".   That's why I changed the name. Do you think there is another way I could avoid this error?

File: core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala
Comment (huxihx) [reply core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: Sorry, my mistake. Do you think it's okay to add props.put("producer.type", "sync") before MirrorMakerProducer?

File: core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala
Comment (ijuma) [reply core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala]: Yeah, that should fail without your fix, I think. If we wanted to make the test even better, we'd produce to `another_topic` and `yet_another_topic` and check that the consumer only consumes from the two that match.

But the test as it is verifies the fix, so I won't block the merge if you don't feel like extending the test. :)

File: core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala
Comment (ijuma) [reply core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: Ah, that makes sense because they both have the same package (it's just the source folder that differs). In that case, I suggest `MirrorMakerIntegrationTest` for this one so the difference is obvious.

File: core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala
Comment (ijuma) [reply core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala]: Sounds good to me.

**Comment by huxihx** [comment]
@ijuma Thanks for your great patience. Please review it again after addressing your comments. Thanks.

File: core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala]: Hmm, this doesn't seem right. Why do we swallow this exception?

File: core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala]: We should use `TestUtils.waitUntilTrue` instead of this.

File: core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala]: We shouldn't have this. Unintentional?

File: core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala:31]: Did it not work to use `classOf[ByteArraySerializer]` instead of hardcoding the String?

**Review by ijuma — CHANGES_REQUESTED** [review]
Thanks for the updates, a few more comments.

File: core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala
Comment (huxihx) [reply core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala:31]: Seems MirrorMakerProducer is not a parameterized class which cannot use type parameters during construction.

File: core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala
Comment (huxihx) [reply core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala]: I always forget to remove them before committing. Sorry, my mistake.

File: core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala
Comment (amethystic) [inline core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala]: `val mirrorMakerConsumer = new MirrorMakerNewConsumer(consumer, None, whitelist)
    mirrorMakerConsumer.init()
    try {
      TestUtils.waitUntilTrue(() => {
        try {
          val data = mirrorMakerConsumer.receive()
          data.topic == topic && new String(data.value) == msg
        } catch {
          case _: ConsumerTimeoutException => false// swallow it
        }
      }, "MirrorMaker consumer should be able to subscribe the correct topic and read the correct message.")
    } finally {
      consumer.close()
    }`
Is that okay to you? Thanks.

Comment (huxihx) [reply core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala]: `val mirrorMakerConsumer = new MirrorMakerNewConsumer(consumer, None, whitelist)

mirrorMakerConsumer.init()

 try {
      TestUtils.waitUntilTrue(() => {
        try {
          val data = mirrorMakerConsumer.receive()
          data.topic == topic && new String(data.value) == msg
        } catch {
          case _: ConsumerTimeoutException => false
        }
      }, "MirrorMaker consumer should be able to subscribe the correct topic and read the correct message.")
    } finally {
      consumer.close()
}`
Is that okay to you then? Thanks.

File: core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala
Comment (ijuma) [reply core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala:31]: That's not what I was suggesting. That's OK, I'll do it before merging.

**Comment by ijuma** [comment]
Thanks for the updates, LGTM with a few minor changes that I will do before merging to trunk: moved upgrade notes to the right place and reworded them a little, added license comment to test, minor style improvements.

**Comment by huxihx** [comment]
Thanks for your great patience. I do learn a lot from this whole review cycle.

**Pull request closed without merging**
